packages/react-pages/src/node/dynamic-modules/UpdateBuffer.ts

```diff
@@ -0,0 +1,114 @@
+import { EventEmitter } from 'events'
+import { debounce } from 'mini-debounce'
+
+/**
+ * Types of page data updates.
+ *
+ * add:
+ *  A new page is added.
+ *  The page list module will be updated.
+ * update:
+ *  A page is updated.
+ *  The page list module will be updated if it is static data change
+ *  The page data module will be updated if it is runtime data change
+ * delete:
+ *  A page is deleted.
+ *  The page list module will be updated.
+ *  Buffered update of the deleted page will be cancled.
+ */
+type Update =
+  | {
+      type: 'add' | 'delete'
+      pageId: string
+    }
+  | {
+      type: 'update'
+      pageId: string
+      dataType: 'runtime' | 'static'
+    }
+
+export type ScheduleUpdate = (update: Update) => void
+
+/**
+ * Buffer page data updates.
+ * Can flush a batch of updates together
+ * and cancle unnecessary updates
+ */
+export class UpdateBuffer extends EventEmitter {
+  /**
+   * which pages should be updated
+   */
+  private pageUpdateBuffer = new Set<string>()
+
+  /**
+   * whether the page list should be updated
+   */
+  private pageListUpdateBuffer = false
+
+  private scheduleFlush: () => void
+
+  constructor() {
+    super()
+    this.scheduleFlush = debounce(() => {
+      if (this.pageUpdateBuffer.size > 0) {
+        const updates = [...this.pageUpdateBuffer.values()]
+        this.emit('page', updates)
+        this.pageUpdateBuffer.clear()
+      }
+
+      if (this.pageListUpdateBuffer) {
+        // if this.pageUpdateBuffer.size >= 0
+        // the page update will automatically update the page list
+        // (because the whole import chain will update)
+        // so we don't need to trigger page list update in that case
+        if (this.pageUpdateBuffer.size === 0) this.emit('page-list')
```

Comment:
`this.pageUpdateBuffer.size === 0` is always true, since you clear the `pageUpdateBuffer` a few lines before that

Reply:
Good catch. Fixed in https://github.com/vitejs/vite-plugin-react-pages/pull/19/commits/a9689e2b0e1bd7e629b839935bd7236206671054